pkg/controller/jobs/pod/pod_controller.go

```diff
@@ -229,6 +264,276 @@ func (p *Pod) Skip() bool {
+		if err != nil {
+			return nil, fmt.Errorf("failed to extract '%s' annotation from the pod '%s': %w",
+				GroupTotalCountAnnotation,
+				podInGroup.GetName(),
+				err)
+		}
+		if tc != groupTotalCount {
+			// ToDo: Add a warning event
+			return nil, fmt.Errorf("pods '%s' and '%s' has different '%s' values: %d!=%d",
+				p.GetName(), podInGroup.GetName(),
+				GroupTotalCountAnnotation,
+				groupTotalCount, tc)
+		}
+
+		roleHash, ok := podInGroup.Annotations[RoleHashAnnotation]
+		if !ok {
+			roleHash, err = getRoleHash(fromObject(&podInGroup))
+		}
+
+		roleHash = roleHash[:8]
+
+		podRoleFound := false
+		for psi := range resultPodSets {
+			if resultPodSets[psi].Name == roleHash {
+				podRoleFound = true
+				resultPodSets[psi].Count++
+			}
+		}
+
+		if !podRoleFound {
+			podSet := fromObject(&podInGroup).PodSets()
+			podSet[0].Name = roleHash
+
+			resultPodSets = append(resultPodSets, podSet[0])
+		}
+	}
+
+	return resultPodSets, nil
+}
+
+func (p *Pod) ConstructWorkload(ctx context.Context, c client.Client, r record.EventRecorder) (*kueue.Workload, error) {
+	log := ctrl.LoggerFrom(ctx)
+	object := p.Object()
+
+	var (
+		podSets     []kueue.PodSet
+		podsInGroup corev1.PodList
+	)
+
+	if p.GetGroupName() == "" {
+		podSets = p.PodSets()
+	} else {
+		if err := c.List(ctx, &podsInGroup, client.MatchingLabels{
+			GroupNameLabel: p.GetGroupName(),
+		}); err != nil {
+			return nil, err
+		}
+
+		groupTotalCount, err := p.GetGroupTotalCount()
+		if err != nil {
+			return nil, err
+		}
+
+		if len(podsInGroup.Items) != groupTotalCount {
+			r.Eventf(object, corev1.EventTypeWarning, "ErrWorkloadCompose", errMsgIncorrectTotalGroupCount)
+			return nil, fmt.Errorf(errMsgIncorrectTotalGroupCount)
+		}
+
+		podSets, err = p.constructGroupPodSets(podsInGroup)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	wl := &kueue.Workload{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace:  p.GetNamespace(),
+			Labels:     map[string]string{},
+			Finalizers: []string{kueue.ResourceInUseFinalizerName},
+		},
+		Spec: kueue.WorkloadSpec{
+			PodSets:   podSets,
+			QueueName: jobframework.QueueName(p),
+		},
+	}
+
+	if p.GetGroupName() == "" {
+		wl.Name = jobframework.GetWorkloadNameForOwnerWithGVK(p.GetName(), p.GVK())
+	} else {
+		wl.Name = p.GetGroupName()
+	}
+
+	if p.GetGroupName() == "" {
+		jobUid := string(p.Object().GetUID())
+		if errs := validation.IsValidLabelValue(jobUid); len(errs) == 0 {
+			wl.Labels[controllerconsts.JobUIDLabel] = jobUid
+		} else {
+			log.V(2).Info(
+				"Validation of the owner job UID label has failed. Creating workload without the label.",
+				"ValidationErrors", errs,
+				"LabelValue", jobUid,
+			)
+		}
+
+		// add the owner ref
+		if err := controllerutil.SetControllerReference(p.Object(), wl, c.Scheme()); err != nil {
+			return nil, err
+		}
+	}
+
+	for _, pod := range podsInGroup.Items {
+		if err := controllerutil.SetOwnerReference(&pod, wl, c.Scheme()); err != nil {
```

Comment:
Are all pods owners? When is the workload deleted? when all the owners are deleted?